pkg/flow/internal/controller/config.go

```diff
@@ -34,3 +37,127 @@ func NewConfigNode(block *ast.BlockStmt, globals ComponentGlobals, isInModule bo
 		return nil, diags
 	}
 }
+
+// ConfigNodeMap represents the config BlockNodes in their explicit types.
+// This is helpful when validating node conditions specific to config node
+// types.
+type ConfigNodeMap struct {
+	logging     *LoggingConfigNode
+	tracing     *TracingConfigNode
+	argumentMap map[string]*ArgumentConfigNode
+	exportMap   map[string]*ExportConfigNode
+}
+
+// NewConfigNodeMap will create an initial ConfigNodeMap. Append must be called
+// to populate NewConfigNodeMap.
+func NewConfigNodeMap() *ConfigNodeMap {
+	return &ConfigNodeMap{
+		logging:     nil,
+		tracing:     nil,
+		argumentMap: map[string]*ArgumentConfigNode{},
+		exportMap:   map[string]*ExportConfigNode{},
+	}
+}
+
+// Append will add a config node to the ConfigNodeMap. This will overwrite
+// values on the ConfigNodeMap that are matched and previously set.
+func (nodeMap *ConfigNodeMap) Append(configNode BlockNode) diag.Diagnostics {
+	var diags diag.Diagnostics
+
+	switch n := configNode.(type) {
+	case *ArgumentConfigNode:
+		nodeMap.argumentMap[n.Label()] = n
+	case *ExportConfigNode:
+		nodeMap.exportMap[n.Label()] = n
+	case *LoggingConfigNode:
+		nodeMap.logging = n
+	case *TracingConfigNode:
+		nodeMap.tracing = n
+	default:
+		diags.Add(diag.Diagnostic{
+			Severity: diag.SeverityLevelError,
+			Message:  fmt.Sprintf("unsupported config node type found %q", n.Block().Name),
+			StartPos: ast.StartPos(n.Block()).Position(),
+			EndPos:   ast.EndPos(n.Block()).Position(),
+		})
+	}
+
+	return diags
+}
+
+// Validate wraps all validators for ConfigNodeMap.
+func (nodeMap *ConfigNodeMap) Validate(isInModule bool, args map[string]any) diag.Diagnostics {
+	var diags diag.Diagnostics
+
+	newDiags := nodeMap.ValidateModuleConstraints(isInModule)
+	diags = append(diags, newDiags...)
+
+	newDiags = nodeMap.ValidateUnsupportedArguments(args)
+	diags = append(diags, newDiags...)
+
+	return diags
+}
+
+// ValidateModuleConstraints will make sure config blocks with module
+// constraints get followed.
+func (nodeMap *ConfigNodeMap) ValidateModuleConstraints(isInModule bool) diag.Diagnostics {
+	var diags diag.Diagnostics
+
+	if isInModule {
+		if nodeMap.logging != nil {
+			diags.Add(diag.Diagnostic{
+				Severity: diag.SeverityLevelError,
+				Message:  "logging block not allowed inside a module",
+				StartPos: ast.StartPos(nodeMap.logging.Block()).Position(),
+				EndPos:   ast.EndPos(nodeMap.logging.Block()).Position(),
+			})
+		}
+
+		if nodeMap.tracing != nil {
+			diags.Add(diag.Diagnostic{
+				Severity: diag.SeverityLevelError,
+				Message:  "tracing block not allowed inside a module",
+				StartPos: ast.StartPos(nodeMap.tracing.Block()).Position(),
+				EndPos:   ast.EndPos(nodeMap.tracing.Block()).Position(),
+			})
+		}
```

Comment:
Thoughts on adding a return here and allowing us to get rid of the else/indentation?